### PR TITLE
Supporting TCRS configuration through gridset selector

### PR DESCRIPTION
**Pull request by dromagnoli**
# Checklist

- [ ] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [ ] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [ ] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by dromagnoli** [comment]
Code depends on this GeoTools PR adding PROJ formatting support: https://github.com/geosolutions-it/geotools/pull/84

MapML XML/XSD has been updated to modify the projection elements.
It was tied to ProjType which is an enum only supporting the 4 Built-in Projections (WGS84,CBMTILE,APSTILE,OSMTILE).
I have created a WrappingProjType which encapsulates the ProjType and modified projection elements of xsd/xml to be simple String instead of ProjType type. Please, let me know if that's ok.

Side Question on the next task related to Projection support:
SW.4.1.19 requires allowing GeoServer admin to set the default projection MapML URL. 
At the moment I think is it already supported.
Admin can specify default Projection during layer configuration, by specifying a MapML:CRS as Declared CRS and select Reproject Native to Declared.
image

The generated LayerPreview will use that CRS. 
http://localhost:8080/geoserver/topp/wms?service=WMS&version=1.1.0&request=GetMap&layers=topp%3Astates&bbox=-2154706.877127081%2C2763762.2272634404%2C3844204.014332743%2C5981537.174078367&width=768&height=411&srs=MapML%3AUTM14WGS84Quad&styles=&format=text%2Fhtml%3B%20subtype%3Dmapml

image
image

Is it what is requested? 
Or do we actually need a different feature where we want the possibility to specify a Declared CRS and specify a different Default output CRS?

**Comment by prushforth** [comment]
> MapML XML/XSD has been updated to modify the projection elements.
> It was tied to ProjType which is an enum only supporting the 4 Built-in Projections (WGS84,CBMTILE,APSTILE,OSMTILE).
> I have created a WrappingProjType which encapsulates the ProjType and modified projection elements of xsd/xml to be simple String instead of ProjType type. Please, let me know if that's ok.

That is ok!  One thing I don't want to have happen is to allow the projection name string to contain `:`, as I don't want users to use for example, "EPSG:4326" as the name of a TCRS, since they are definitely not the same kind of thing.  I noticed that the data type used in the rnc was `string` but in the xsd it's `anyURI` which might lead to trouble, but I suppose even a plain string without the `:` can be considered a URI.

> Side Question on the next task related to Projection support:
> SW.4.1.19 requires allowing GeoServer admin to set the default projection MapML URL.
> At the moment I think is it already supported.

You are correct.

> Admin can specify default Projection during layer configuration, by specifying a MapML:CRS as Declared CRS and select Reproject Native to Declared.

This is done, thanks!

**Comment by dromagnoli** [comment]
> One thing I don't want to have happen is to allow the projection name string to contain `:` 

This is already addressed, there is also a note in the GridSet to TCRS Selector documentation that reports that:
https://github.com/geosolutions-it/geoserver/pull/375/files#diff-3bb5273b2d308af6af5f3b5f39e968a4307b50fe4084adf3e6dda3f2c8935fe8R73

Quoting:
**_Gridsets containing ":" character in the name won't be listed_**

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java [hunk 1]
```diff
@@ -506,30 +504,27 @@ public static List<Link> alternateProjections(
...
-                });
-
+        for (String proj : projections) {
+            Link l = new Link();
+            TiledCRSParams projection = TiledCRSConstants.lookupTCRSParams(proj);
+            boolean addMe = false;
+            l.setProjection(projection.getName());
+            addMe = true;
+
+            if (!addMe) continue;
```
Comment (aaime) [inline src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java:514]: Uh? How does one even get "addMe" to be false there? Sounds like a debugging code bit left around.

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java [hunk 2]
```diff
@@ -506,30 +504,27 @@ public static List<Link> alternateProjections(
...
-                    l.setHref(
-                            ResponseUtils.urlDecode(
-                                    ResponseUtils.buildURL(
-                                            base, path, kvp, URLMangler.URLType.SERVICE)));
-                    links.add(l);
-                });
-
+        for (String proj : projections) {
+            Link l = new Link();
+            TiledCRSParams projection = TiledCRSConstants.lookupTCRSParams(proj);
```
Comment (aaime) [inline src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java:509]: I'm confused by the overall change.. seems just a switch between streams and a foreach, or did I miss something?

File: src/extension/mapml/src/test/java/org/geoserver/mapml/tcrs/CustomTiledCRSTest.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                        tileHeight, // Tile height (256 pixels)
+                        yCoordinateFirst // Whether Y-coordinate comes first (false)
+                        );
+
+        return gridSet;
+    }
+
+    @Test
+    public void testGridsetToTiledCRS() throws Exception {
+        CoordinateReferenceSystem crs = CRS.decode("MapML:UTM31WGS84Quad");
```
Comment (aaime) [inline src/extension/mapml/src/test/java/org/geoserver/mapml/tcrs/CustomTiledCRSTest.java:122]: Good. I would also check that:
* The code shows up in "CRS.getSupportedCodes()
* A GUI test that adds a built-in valid griset as a valid TRCS, and after the submission, the CRS class can find it

**Review by aaime — CHANGES_REQUESTED** [review]
Overall looks good, see comments below

File: src/extension/mapml/src/main/java/org/geoserver/mapml/tcrs/TiledCRSInfo.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package org.geoserver.mapml.tcrs;
+
+import java.io.Serializable;
+import java.util.Set;
+
+public interface TiledCRSInfo extends Cloneable, Serializable {
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/tcrs/TiledCRSInfo.java:6]: The class is not used. I have committed that by mistake before a refactoring. Removing it

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java (hunk 2 above)
Comment (dromagnoli) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java:509]: Restored. Not sure why it changed (probably as part of some intermediate development).

File: src/extension/mapml/src/main/java/org/geoserver/mapml/gwc/gridset/MapMLGridsets.java [hunk 5]
```diff
@@ -8,32 +8,84 @@
...
+
     private final GridSet WGS84;
     private final GridSet OSMTILE;
     private final GridSet CBMTILE;
     private final GridSet APSTILE;
-    @Autowired private GWC gwc = GWC.get();
 
     /** */
     public MapMLGridsets() {
+        gwc = GeoServerExtensions.bean(GWC.class);
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/gwc/gridset/MapMLGridsets.java:88]: Fixed.

File: src/extension/mapml/src/main/java/org/geoserver/mapml/tcrs/TiledCRSConstants.java [hunk 6]
```diff
@@ -279,32 +291,112 @@ public static String getSupportedOutputCRS(String proj) {
...
+                    String code = param.getCode();
+                    tiledCRSDefinitions.put(name.toUpperCase(), param);
+                    tiledCRSDefinitions.put(code, param);
+                    tiledCRSBySrsName.put(code, param);
+                    TiledCRS tiledCRS = new TiledCRS(name.toUpperCase(), param);
+                    TILED_CRS.put(code, tiledCRS);
+                    TILED_CRS.put(name, tiledCRS);
+                    TILED_CRS.put(("MAPML:" + name).toUpperCase(), tiledCRS);
+                    if (code.toUpperCase().startsWith("EPSG:")) {
+                        String urnEPSGcode = code.replace("EPSG:", "URN:OGC:DEF:CRS:EPSG::");
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/tcrs/TiledCRSConstants.java:391]: Fixed.  using SrsSyntax.OGC_URN.getSRS(code).

File: src/extension/mapml/src/test/java/org/geoserver/mapml/tcrs/CustomTiledCRSTest.java (hunk 3 above)
Comment (dromagnoli) [reply src/extension/mapml/src/test/java/org/geoserver/mapml/tcrs/CustomTiledCRSTest.java:122]: Tests Added.

**Comment by dromagnoli** [comment]
@aaime.
All the feedbacks have been addressed, please, let me know.

**Comment by aaime** [comment]
Go for the official repo!

**Comment by aaime** [comment]
@prushforth fix for axis order merged here: https://github.com/geoserver/geoserver/pull/8143

**Pull request closed without merging**